### PR TITLE
lib-user: Add certificate page

### PR DESCRIPTION
**Pull request by mcbouslog**
## Package
- app-root
- lib-react-components
- lib-user

## Describe your changes
Figma:
- pages: https://www.figma.com/design/qbqbmR3t5XV6eKcpuRj7mG/Group-Stats?node-id=16-110&t=evrDTlBn4g63ZzlS-4
- certificates: https://www.figma.com/design/qbqbmR3t5XV6eKcpuRj7mG/Group-Stats?node-id=1827-6388&t=evrDTlBn4g63ZzlS-4

Packages:
- app-root
  - add certificate page
  - add UserStatsContext (to share user stats state - dateRange and project - between user stats page and certificate page)
  - add stats layout.js for UserStatsContextProvider (applies to `/stats` and `/stats/certificate`)
  - add userStatsDateRange hook (initially state, but added localStorage to persist across browser tabs)
  - add userStatsProject hook (same as useStatsDateRange)
- lib-react-components
  - refactor ZooFooter and ZooHeader with `display: none` when printing (`@media print`)
- lib-user
  - add CertificateContainer and Certificate components for user volunteer certificate display and generating (printing to PDF or printer)
  - add Certificate stories
  - add certificate page to dev app
  - refactor UserStatsContainer for selectedDateRange and selectedProject as props
  - refactor CertificateContainer for selectedDateRange and selectedProject as props 

## How to Review
_Helpful explanations that will make your reviewer happy:_
- What Zooniverse project should my reviewer use to review UX? n/a
- What user actions should my reviewer step through to review this PR?
  - from app-root go to your user stats page (i.e. https://local.zooniverse.org:3000/users/markb-panoptes/stats)
  - confirm user stats page projects and date range selects work as expected
  - click (and right click to new tab) the Generate Volunteer Certificate link to go to your certificate page
  - confirm certificate page reflects user stats project and date range selections
  - click Generate Certificate button/link to open browser print dialog to print certificate to PDF or printer
- Which storybook stories should be reviewed?
  - certificate page: http://localhost:6008/?path=/story/components-certificate--default
- Are there plans for follow up PR’s to further fix this bug or develop this feature? see linked GitHub project

# Checklist
_PR Creator - Please cater the checklist to fit the review needed for your code changes._
_PR Reviewer - Use the checklist during your review. Each point should be checkmarked or discussed before PR approval._

## General
- [ ] Tests are passing locally and on Github
- [ ] Documentation is up to date and changelog has been updated if appropriate
- [ ] You can `yarn panic && yarn bootstrap` or `docker-compose up --build` and FEM works as expected
- [ ] FEM works in all major desktop browsers: Firefox, Chrome, Edge, Safari (Use Browserstack account as needed)
- [ ] FEM works in a mobile browser

## General UX
Example Staging Project: [i-fancy-cats](https://local.zooniverse.org:3000/projects/brooke/i-fancy-cats)
- [ ] All pages of a FEM project load: Home Page, Classify Page, and About Pages
- [ ] Can submit a classification
- [ ] Can sign-in and sign-out
- [ ] The component is accessible
  - Can be used with a screen reader [BBC guide to VoiceOver](https://bbc.github.io/accessibility-news-and-you/assistive-technology/testing-steps/voiceover-mac.html)
  - Can be used from the keyboard [WebAIM guide to keyboard testing](https://webaim.org/techniques/keyboard/#testing)
  - It is passing accessibility checks in the storybook

## New Feature
- [ ] The PR creator has listed user actions to use when testing the new feature
- [ ] Unit tests are included for the new feature
- [ ] A storybook story has been created or updated
  - Example of SlideTutorial [component](https://github.com/zooniverse/front-end-monorepo/blob/master/packages/lib-classifier/src/components/Classifier/components/SlideTutorial/SlideTutorial.js) with docgen comments, and its [story](https://zooniverse.github.io/front-end-monorepo/@zooniverse/classifier/index.html?path=/docs/other-slidetutorial--default)

### CODE REVIEW

File: packages/lib-user/src/components/shared/MainContent/MainContent.js [hunk 1]
```diff
@@ -130,7 +131,9 @@ function MainContent({
           gap='16px'
           justify='end'
         >
-          <button type='button' onClick={() => alert('Coming soon!')}>Generate Volunteer Certificate</button>
+          <Link href={`/users/${source.login}/stats/certificate`}>
```
Comment (mcbouslog) [inline packages/lib-user/src/components/shared/MainContent/MainContent.js:134]: I'll inquire about styling; should it be a (link that looks like a) button per Figma or a link similar to other content section links?

**Review by shaunanoordin — APPROVED** [review]
## PR Review

Package: app-root, lib-user, (minor: lib-react-components)

This PR adds the much-requested **volunteer participation certificate** to the _user stats page_ on app-root. Hurrah! 🎉 

image

Features:
- Volunteers generate their certificate by...
  - Logging in and going to their user stats page, at the path `/users/USERNAME/stats` [(example)](https://local.zooniverse.org:3000/users/darkeshard/stats/certificate)
  - Selecting 1. a project (or all projects) and 2. the date range they're interested in.
  - Clicking on the "Generate Volunteer Certificate" link.
- The volunteer participation certificate lists: 1. user's display name, 2. number of hours (as calculated by number of classifications, IIRC?) spent on the selected project(s) within the selected date range, 3. the selected project number (or number of projects), 4. the selected date range, along with 5. some static Zooniverse information
  - The certificate itself exists on the path `/users/USERNAME/stats/certificate`... (see dev notes below)
  - ...and uses input values that's set on a different page, and partially persisted via local storage. (again, see dev notes)
  - A "generate certificate" link is available on the certificate page, which triggers a print action (which neatly hides the footer and header. Nice!)
- A user can only view their own stats page 
  - Attempting to view the stats URL or certificate URL for another user, OR attempting to view either URL when not logged in, will cause the page to simply display the error message: `Not authorized`
- 🦆 Quirk: admins in admin mode can view the stats page of any user, but can't view the certificate of another user.

Changed behaviour:
- If a user **prints** _any_ Zooniverse page running on FEM, the **Zooniverse website Header** and **Footer** will now ALWAYS be hidden.
  - ❓ @mcbouslog and/or @seanmiller26 - I wanted to double check that this is intended new behaviour moving forward. On one hand, it's a pretty wide-reaching global change. On the other hand, I'm not sure who ever prints websites. (This change does not affect screenshots.)

Code looks good, and functionality of the certificate page checks out. This PR looks good to merge, though I have a bunch of (non-blocking) dev notes that I want to share.

### Dev Notes

Cleaner code:
- At the moment, the _inputs_ to the certificate page is actually set in the separate user stats page. The input values of selected project(s) and selected date range are passed via app context (and local storage, in case users type in the certificate URL directly).
- While this definitely works, it could potentially make debugging a bit messy, as the use of local storage may mean we're not always 100% sure what inputs went in to create a specific output. I think the React language for this is, "the component isn't pure"? or "the function isn't pure"?
- Compare this with, e.g. if the certificate page pulls its inputs explicitly from the URL, such as `https://local.zooniverse.org:3000/users/darkeshard/stats/certificate?project=all&daterange=last7days`.
- No action is required for this note - this is just me raising an eyebrow a bit like 🤨 "hmmm!"

UI/UX:
- On the [user stats page,](https://local.zooniverse.org:3000/users/darkeshard/stats) the link to the certificate page says "Generate Volunteer Certificate"
- On the [certificate page,](https://local.zooniverse.org:3000/users/darkeshard/stats/certificate) the link to trigger the print action says "generate certificate"
- 🤔 Thought/minor suggestion: maybe that second link should say "print certificate"?

Potential user shenanigans:
- I'm thinking of several ways that a volunteer might _not_ use the volunteer certificate page correctly, and I'm wondering if we should preempt this with any sort of guidance. I'm not sure how much we need to consider these actually, so I'm going to need to defer to @seanmiller26 about 'em.
- Scenario 1: a volunteer, not realising they need to print the certificate, instead shares [the URL of their certificate](https://local.zooniverse.org:3000/users/darkeshard/stats/certificate) with their teacher or supervisor. The teacher/supervisor, obviously, can't access the page.
- Scenario 2: [minor] a volunteer bookmarks the certificate page URL. However, the page doesn't always display what they expect since the inputs would change. (e.g. they bookmarked it with project = I Fancy Cats and date range = last 30 days, but when they return, it's now reset to project = all and date range = last 7 days)
- Scenario 3: date ranges also means that sometimes a volunteer can never generate the same certificate twice. e.g. "Hey, yesterday I went to the certificate page and it says I did 18 hours, but when I looked at the page today, it says I only did 15!" (Ans: user didn't realise the 7 day window just moved.)

### Status

LGTM! 👍 I would still like to double-confirm the new behaviour changes to ZooFooter and ZooHeader are intended, but otherwise this PR is good to go.

**Comment by mcbouslog** [comment]
Thank you @shaunanoordin ! I'll respond to all of the very helpful points noted, but wanted to note I 100% agree regarding moving the stats date range and project state to the url. Which makes this PR's changes storing state in localStorage overkill (for a temporary solution), but I'm going to leave the localStorage for now for internal testing.

I also think it would be preferable to have the project and date range selects on the certificate page. I mentioned it in related meetings and it was determined to be a "nice to have," but after using the certificate page myself and based on full release timing I'm going to advocate increasing that functionality's priority.

**Comment by seanmiller26** [comment]
> If a user prints any Zooniverse page running on FEM, the Zooniverse website Header and Footer will now ALWAYS be hidden.

Could someone post a quick video of this in action? Specifically going from the stats page to the Certificate page, then clicking to print the page. Thanks!

**Comment by mcbouslog** [comment]
@shaunanoordin and @seanmiller26 - hiding the main header and footer for all pages when printing was discussed during a stats/homepages meeting and it was determined this would be the initial solution. Adding a hide state and passing it around and toggling it can be added if there's related negative feedback.

@seanmiller26 - this PR will likely merge today at which point you can interact with it and we can iterate on if that's ok?

File: packages/lib-user/src/components/shared/MainContent/MainContent.js (hunk 1 above)
Comment (mcbouslog) [reply packages/lib-user/src/components/shared/MainContent/MainContent.js]: Done in 4908653

File: packages/lib-user/src/components/Certificate/Certificate.stories.js
[the remark this answers is not in the document]
Comment (mcbouslog) [reply packages/lib-user/src/components/Certificate/Certificate.stories.js:1]: I think once this PR is deployed to fe-root.preview I can update the URL to something permanent.

File: packages/lib-user/dev/components/App/App.js [hunk 2]
```diff
@@ -27,6 +28,8 @@ function App({
   const [activeIndex, setActiveIndex] = useState(-1)
   const [dark, setDarkTheme] = useState(false)
   const [loading, setLoading] = useState(false)
+  const [selectedDateRange, setSelectedDateRange] = useState('Last7Days')
+  const [selectedProject, setSelectedProject] = useState('AllProjects')
```
[the remark this answers is not in the document]
Comment (mcbouslog) [reply packages/lib-user/dev/components/App/App.js:32]: Absolutely! I'll revisit when I refactor to move state to URL query params.

**Comment by mcbouslog** [comment]
A brief summary per review notes...

Items that will be addressed in a subsequent PR:
- refactor stats date range and project state to url query param
- (eventually, priority TBD) add stats date range and project selects to certificate page
- (eventually, priority TBD) admin access

Items to be monitored:
- on print hiding general Zooniverse header and footer for all pages (implemented specifically for printing user certificate), if negative feedback, or regardless, could implement related conditional state/context

**Pull request merged**